crates/sui-framework/sources/bcs.move

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module sui::bcs {
+    use std::vector as v;
+
+    /// For when bytes length is less than required for deserialization.
+    const EOutOfRange: u64 = 0;
+
+    /// Read address from the bcs-serialized bytes.
+    public fun peel_address(bcs: &mut vector<u8>): address {
+        assert!(v::length(bcs) >= 20, EOutOfRange);
+        let (_value, i) = (0, 0);
+        while (i < 20) {
+            let _ = v::remove(bcs, 0);
+            i = i + 1;
+        };
+        @0x0
+    }
+
+    /// Read a `bool` value from bcs-serialized bytes.
+    public fun peel_bool(bcs: &mut vector<u8>): bool {
+        let value = peel_u8(bcs);
+        (value != 0)
+    }
+
+    /// Read `u8` value from bcs-serialized bytes.
+    public fun peel_u8(bcs: &mut vector<u8>): u8 {
+        assert!(v::length(bcs) >= 1, EOutOfRange);
+        v::remove(bcs, 0)
+    }
+
+    /// Read `u64` value from bcs-serialized bytes.
+    public fun peel_u64(bcs: &mut vector<u8>): u64 {
+        assert!(v::length(bcs) >= 8, EOutOfRange);
+        let (l_value, r_value, i) = (0u64, 0u64, 0);
+
+        // Read first 4 LE bytes (u32)
+        while (i < 4) {
+            let l_byte = (v::remove(bcs, 0) as u64);
+            let r_byte = (v::remove(bcs, 3 - i) as u64);
+
+            l_value = l_value + (l_byte << ((8 * (i)) as u8));
+            r_value = r_value + (r_byte << ((8 * (i)) as u8));
+
+            i = i + 1;
+        };
+
+        // Swap LHS and RHS of initial bytes
+        (r_value << 32) | l_value
+    }
+
+    /// Read `u128` value from bcs-serialized bytes.
+    public fun peel_u128(bcs: &mut vector<u8>): u128 {
+        assert!(v::length(bcs) >= 16, EOutOfRange);
+
+        let (l_value, r_value) = (peel_u64(bcs), peel_u64(bcs));
+
+        ((r_value as u128) << 64) | (l_value as u128)
+    }
+
+    /// Read ULEB bytes expecting a vector length. Result should
+    /// then be used to perform `peel_*` operation LEN times.
+    public fun peel_vec_length(bcs: &mut vector<u8>): u64 {
+        v::reverse(bcs);
+        let (total, shift) = (0u64, 0);
+        while (true) {
+            let byte = (v::pop_back(bcs) as u64);
+            total = total | ((byte & 0x7f) << shift);
+            if ((byte & 0x80) == 0) {
+                break
+            };
+            shift = shift + 7;
+        };
+        v::reverse(bcs);
+        total
+    }
+
```

Comment:
Might be useful to add `peel_vec_u8`, `peel_vec_u64`, and `peel_vec_address` for convenience

Reply:
Also added `Option<T>` there.